sql/in_mem_table/inmem_table_data.go

```diff
@@ -34,29 +35,50 @@ import (
 // of an integrator's domain, while requiring alteration. This is unlike the information_schema tables, which do not
 // allow any kind of direct alteration.
 
-// InMemTableDataKey represents a key that will be matched in order to retrieve a row. All implementations of this
-// interface must NOT be a pointer, as pointers have different rules for determining equality.
-type InMemTableDataKey interface {
-	// AssignValues returns a new key with the given values. These values will be in the same order as their represented
-	// columns (as returned by RepresentedColumns). The new key's type is expected to match the calling key's type.
-	AssignValues(vals ...interface{}) (InMemTableDataKey, error)
-	// RepresentedColumns returns the index positions of the columns that represent this key, based on the table's schema.
-	RepresentedColumns() []uint16
+// Key represents a key that will be matched in order to retrieve a row. All implementations of this interface must NOT
+// be a pointer, as pointers have different rules for determining equality.
+type Key interface {
+	// KeyFromEntry returns a new key from the given entry. The new key's type is expected to match the calling key's type.
+	KeyFromEntry(ctx *sql.Context, entry Entry) (Key, error)
+	// KeyFromRow returns a new key from the given sql.Row. The given row will match the schema of the parent table. The
+	// new key's type is expected to match the calling key's type.
+	KeyFromRow(ctx *sql.Context, row sql.Row) (Key, error)
 }
 
-// InMemTableData is used for in-memory tables to store their row data in a way that may be quickly retrieved for queries.
-type InMemTableData struct {
-	sch                 sql.Schema
+// Entry is an entry in Data. It handles conversions to and from a sql.Row, so that the underlying
+// representation does not have to be a sql.Row. As the sql.Row interface is only important in the context of table
+// operations, the data type stored may be of any type as long as it can convert between itself and sql.Row, as well as
+// testing equality against other entries. All rows will match the schema of the parent table.
+type Entry interface {
+	// NewFromRow takes the given row and returns a new instance of the Entry containing the properties of the given row.
+	NewFromRow(ctx *sql.Context, row sql.Row) (Entry, error)
+	// UpdateFromRow uses the given row to return a new Entry that is based on the calling Entry. This means that any
+	// fields that do not have a direct mapping to or from a sql.Row should be preserved on the new Entry.
+	UpdateFromRow(ctx *sql.Context, row sql.Row) (Entry, error)
+	// ToRow returns this Entry as a sql.Row.
+	ToRow(ctx *sql.Context) sql.Row
+	// Equals returns whether the calling entry is equivalent to the given entry. Equivalence should only be determined
+	// on values that have a direct mapping to and from a sql.Row.
+	Equals(ctx *sql.Context, otherEntry Entry) bool
```

Comment:
This seems an odd addition given the warning about pointer types above.

Why not just use == with structs to compare them?

Reply:
So a `Key` is used in the internal map, so it can't be a pointer, otherwise map comparisons will use the pointer address rather than the struct contents. This is done to prevent potential bugs. If we had generics and I could constrain on non-pointer types, then the warning would be unnecessary.

An `Entry`, on the other hand, is used to represent some kind of data. We could technically constrain these to non-pointers as well and do struct comparisons, however entries may hold some "additional" data that cannot be directly manipulated by interacting with via SQL data manipulation statements (`INSERT`, etc.). Perhaps these are hidden columns, but AFAICT MySQL has _some_ data in these internal tables that can't be seen from the outside. Basically this "additional" data mimics that, but it shouldn't affect equality. Like imagine it tracks some autoincrement ID internally, that would cause every entry to always be unique, but that may just be something for internal usage. That's the difference.

Reply:
I updated the comment to reflect this.